azurerm/internal/services/cdn/cdn_endpoint_custom_domain_resource.go

```diff
@@ -0,0 +1,166 @@
+			Delete: schema.DefaultTimeout(20 * time.Hour),
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validate.CdnEndpointCustomDomainName(),
+			},
+
+			"cdn_endpoint_id": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validate.EndpointID,
+			},
+
+			"host_name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				// Ipv6 (with square enclosed), Ipv4, domain name is allowed
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+		},
+	}
+}
+
+func resourceArmCdnEndpointCustomDomainCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*clients.Client).Cdn.CustomDomainsClient
+	ctx, cancel := timeouts.ForCreate(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	name := d.Get("name").(string)
+	epid := d.Get("cdn_endpoint_id").(string)
+
+	cdnEndpointId, err := parse.EndpointID(epid)
+	if err != nil {
+		return fmt.Errorf("parsing CDN Endpoint ID %q: %+v", epid, err)
+	}
+
+	id := parse.NewCustomDomainID(cdnEndpointId.SubscriptionId, cdnEndpointId.ResourceGroup, cdnEndpointId.ProfileName, cdnEndpointId.Name, name)
+
+	existing, err := client.Get(ctx, id.ResourceGroup, id.ProfileName, id.EndpointName, id.Name)
+	if err != nil {
+		if !utils.ResponseWasNotFound(existing.Response) {
+			return fmt.Errorf("checking for existing Cdn Endpoint Custom Domain %q: %+v", id, err)
+		}
+	}
+
+	if existing.ID != nil && *existing.ID != "" {
+		id, err := parse.CustomDomainID(*existing.ID)
+		if err != nil {
+			return err
+		}
+		return tf.ImportAsExistsError("azurerm_cdn_endpoint_custom_domain", id.ID())
+	}
+
+	props := cdn.CustomDomainParameters{
+		CustomDomainPropertiesParameters: &cdn.CustomDomainPropertiesParameters{
+			HostName: utils.String(d.Get("host_name").(string)),
+		},
+	}
+
+	future, err := client.Create(ctx, id.ResourceGroup, id.ProfileName, id.EndpointName, id.Name, props)
+	if err != nil {
+		return fmt.Errorf("creating Cdn Endpoint Custom Domain %q: %+v", id, err)
+	}
+	if err := future.WaitForCompletionRef(ctx, client.Client); err != nil {
+		return fmt.Errorf("waiting for creation of Cdn Endpoint Custom Domain %q: %+v", id, err)
+	}
+
+	d.SetId(id.ID())
+
+	return resourceArmCdnEndpointCustomDomainRead(d, meta)
+}
+
+func resourceArmCdnEndpointCustomDomainRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*clients.Client).Cdn.CustomDomainsClient
+	ctx, cancel := timeouts.ForRead(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	id, err := parse.CustomDomainID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	resp, err := client.Get(ctx, id.ResourceGroup, id.ProfileName, id.EndpointName, id.Name)
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			log.Printf("[DEBUG] Cdn Endpoint Custom Domain %q was not found - removing from state!", id)
```

Comment:
In case the parse got error, it will return error right after parsing. Otherwise, the pointer (`id`) is guaranteed to be non-nil and the `log.Printf` will use the `String()` of `EndpointId` (as it is in the **method set** of `*EndpointId`).

So I don't think we need to deref the `id` here?